iir.py

```diff
@@ -0,0 +1,154 @@
+        self.inp = inp = Array(Signal((w_data, True)) for _ in range(n_channels))
+        self.outp = outp = Array(Signal((w_data, True)) for _ in range(n_channels))
+        # ab registers for all channels and profiles
+        self.ab = ab = Array(
+            Array(
+                Array(Signal((w_coeff, True)) for _ in range(n_channels))
+                for _ in range(n_profiles)
+            )
+            for _ in range(N_COEFF)
+        )
+        self.offset = offset = Array(
+            Array(Signal((w_data, True)) for _ in range(n_channels))
+            for _ in range(n_profiles)
+        )
+        # registers for selected profile for channel
+        self.ch_profile = ch_profile = Array(
+            Signal(max=n_profiles + 1) for _ in range(n_channels)
+        )
+        # output hold signal for each channel
+        self.hold = hold = Array(Signal() for _ in range(n_channels))
+        ###
+        self.xy = xy = Array(
+            Array(
+                Array(Signal((w_data, True)) for _ in range(n_channels))
+                for _ in range(n_profiles)
+            )
+            for _ in range(N_COEFF)
+        )
+        self.y0_clipped = y0_clipped = Signal((w_data, True))
+        # position in profiles call this index
+        self.pp = pp = Signal(max=n_profiles + 1)
+        self.pc = pc = Signal(max=n_channels + 1)  # position in channels
+        self.busy = busy = Signal()
+        self.step = step = Signal(2)  # computation step
+        self.ch_profile_last_ch = ch_profile_last_ch = Signal(max=n_profiles + 1)
+        self.submodules.dsp = dsp = Dsp()
+        assert w_data <= len(dsp.b)
+        assert w_coeff <= len(dsp.a)
+        shift_c = len(dsp.a) + len(dsp.b) - w_data - (w_data - log2_a0)
+        shift_a = len(dsp.a) - w_coeff
+        shift_b = len(dsp.b) - w_data
+        # +1 from standard sign bit
+        n_sign = len(dsp.p) - len(dsp.a) - len(dsp.b) + w_data - log2_a0 + 1
+        c_rounding_offset = (1 << shift_c - 1) - 1
+        if decoder != None:
+            self.comb += [
+                ab[k][j][i].eq(decoder.get(f"ch{i}_profile{j}_coeff{k}", "write"))
+                for i in range(n_channels)
+                for j in range(n_profiles)
+                for k in range(N_COEFF)
+            ]
+            self.comb += [
+                offset[j][i].eq(decoder.get(f"ch{i}_profile{j}_offset", "write"))
+                for i in range(n_channels)
+                for j in range(n_profiles)
+            ]
+            self.sync += [
+                If(
+                    stb_out,
+                    # bit 0 is the ch enable bit, bit 1 the hold bit
+                    ch_profile[0].eq(decoder.get(f"servo0_cfg", "write") >> 2),
+                    ch_profile[1].eq(decoder.get(f"servo1_cfg", "write") >> 2),
+                    hold[0].eq(decoder.get(f"servo0_cfg", "write") >> 1),
+                    hold[1].eq(decoder.get(f"servo1_cfg", "write") >> 1),
```

Comment:
Better not use shifts here, just bitslice.
```suggestion
                    ch_profile[0].eq(decoder.get(f"servo0_cfg", "write")[2:]),
                    ch_profile[1].eq(decoder.get(f"servo1_cfg", "write")[2:]),
                    hold[0].eq(decoder.get(f"servo0_cfg", "write")[1]),
                    hold[1].eq(decoder.get(f"servo1_cfg", "write")[1]),
```